File: core/matchers/src/main/scala/au/com/dius/pact/matchers/XmlBodyMatcher.scala
package au.com.dius.pact.matchers

import java.util.Collections

import au.com.dius.pact.model._
import au.com.dius.pact.model.matchingrules.MatchingRules
import com.typesafe.scalalogging.StrictLogging

import scala.collection.JavaConverters._
import scala.xml._

class XmlBodyMatcher extends BodyMatcher with StrictLogging {

  override def matchBody(expected: HttpPart, actual: HttpPart, allowUnexpectedKeys: Boolean): java.util.List[BodyMismatch] = {
    (expected.getBody.getState, actual.getBody.getState) match {
      case (OptionalBody.State.MISSING, _) => Collections.emptyList()
      case (OptionalBody.State.NULL, OptionalBody.State.PRESENT) => Collections.singletonList(new BodyMismatch(null, actual.getBody.getValue,
        s"Expected empty body but received '${actual.getBody.getValue}'"))
      case (OptionalBody.State.NULL, _) => Collections.emptyList()
      case (_, OptionalBody.State.MISSING) => Collections.singletonList(new BodyMismatch(expected.getBody.getValue, null,
        s"Expected body '${expected.getBody.getValue}' but was missing"))
      case (OptionalBody.State.EMPTY, OptionalBody.State.EMPTY) => Collections.emptyList()
      case (_, _) => compareNode(Seq("$"), parse(expected.getBody.orElse("")),
        parse(actual.getBody.orElse("")), allowUnexpectedKeys, expected.getMatchingRules).asJava
    }
  }

  def parse(xmlData: String) = {
    if (xmlData.isEmpty) Text("")
    else Utility.trim(XML.loadString(xmlData))
  }

  def appendIndex(path:Seq[String], index:Integer): Seq[String] = {
    path :+ index.toString
  }

  def appendAttribute(path:Seq[String], attribute: String) : Seq[String] = {
    path :+ "@" + attribute
  }

  def mkPathString(path:Seq[String]) = path.mkString(".")


  def compareText(path: Seq[String], expected: Node, actual: Node, allowUnexpectedKeys: Boolean,
                  matchers: MatchingRules): List[BodyMismatch] = {
    val textpath = path :+ "#text"
    val expectedText = expected.child.filter(n => n.isInstanceOf[Text]).map(n => n.text).mkString
    val actualText = actual.child.filter(n => n.isInstanceOf[Text]).map(n => n.text).mkString
    if (Matchers.matcherDefined("body", textpath.asJava, matchers)) {
      logger.debug("compareText: Matcher defined for path " + textpath)
      Matchers.domatch[BodyMismatch](matchers, "body", textpath.asJava, expectedText, actualText, BodyMismatchFactory.INSTANCE).asScala.toList
    } else if (expectedText != actualText) {
      List(new BodyMismatch(expected, actual, s"Expected value '$expectedText' but received '$actualText'", mkPathString(textpath)))
    } else {
      List()
    }
  }

  def compareNode(path: Seq[String], expected: Node, actual: Node, allowUnexpectedKeys: Boolean,
                  matchers: MatchingRules): List[BodyMismatch] = {
    val nodePath = path :+ expected.label
    val mismatches = if (Matchers.matcherDefined("body", nodePath.asJava, matchers)) {
      logger.debug("compareNode: Matcher defined for path " + nodePath)
      Matchers.domatch[BodyMismatch](matchers, "body", nodePath.asJava, expected, actual, BodyMismatchFactory.INSTANCE).asScala.toList
    } else if (actual.label != expected.label) {
        List(new BodyMismatch(expected, actual, s"Expected element ${expected.label} but received ${actual.label}", mkPathString(nodePath)))
    } else {
        List()
    }

    if (mismatches.isEmpty) {
      compareAttributes(nodePath, expected, actual, allowUnexpectedKeys, matchers) ++
        compareChildren(nodePath, expected, actual, allowUnexpectedKeys, matchers) ++
        compareText(nodePath, expected, actual, allowUnexpectedKeys, matchers)
    } else {
      mismatches
    }
  }

  private def compareChildren(path: Seq[String], expected: Node, actual: Node, allowUnexpectedKeys: Boolean,
                           matchers: MatchingRules): List[BodyMismatch] = {
    var expectedChildren = expected.child.filter(n => n.isInstanceOf[Elem])
    val actualChildren = actual.child.filter(n => n.isInstanceOf[Elem])
    val mismatches = if (Matchers.matcherDefined("body", path.asJava, matchers)) {
      if (expectedChildren.nonEmpty) expectedChildren = expectedChildren.padTo(actualChildren.length, expectedChildren.head)
      List()
    } else if (expected.child.isEmpty && actual.child.nonEmpty && !allowUnexpectedKeys) {
        List(new BodyMismatch(expected, actual, s"Expected an empty List but received ${actual.child.mkString(",")}", mkPathString(path)))
    } else if (expected.child.size != actual.child.size) {
        val missingChilds = expected.child.diff(actual.child)
        val result = missingChilds.map(child => new BodyMismatch(expected, actual, s"Expected $child but was missing", mkPathString(path)))
        if (allowUnexpectedKeys && expected.child.size > actual.child.size) {
          result.toList :+ new BodyMismatch(expected, actual,
            s"Expected a List with atleast ${expected.child.size} elements but received ${actual.child.size} elements", mkPathString(path))
        } else if (!allowUnexpectedKeys && expected.child.size != actual.child.size) {
          result.toList :+ new BodyMismatch(expected, actual,
            s"Expected a List with ${expected.child.size} elements but received ${actual.child.size} elements", mkPathString(path))
        } else {
          result.toList
        }
    } else List()

    mismatches ++: expectedChildren
        .zipWithIndex
        .zip(actualChildren)
        .flatMap(x => compareNode(appendIndex(path, x._1._2), x._1._1, x._2, allowUnexpectedKeys, matchers)).toList
  }

  private def compareAttributes(path: Seq[String], expected: Node, actual: Node, allowUnexpectedKeys: Boolean,
                              matchers: MatchingRules): List[BodyMismatch] = {
    val expectedAttrs = expected.attributes.asAttrMap
    val actualAttrs = actual.attributes.asAttrMap

    if (expectedAttrs.isEmpty && actualAttrs.nonEmpty && !allowUnexpectedKeys) {
      List(new BodyMismatch(expected, actual,
        s"Expected a Tag with at least ${expectedAttrs.size} attributes but received ${actual.attributes.size} attributes",
        mkPathString(path)))
    } else {
      val mismatches = if (allowUnexpectedKeys && expectedAttrs.size > actualAttrs.size) {
        List(new BodyMismatch(expected, actual, s"Expected a Tag with at least ${expected.attributes.size} attributes but received ${actual.attributes.size} attributes",
          mkPathString(path)))
      } else if (!allowUnexpectedKeys && expectedAttrs.size != actualAttrs.size) {
        List(new BodyMismatch(expected, actual, s"Expected a Tag with ${expected.attributes.size} attributes but received ${actual.attributes.size} attributes",
          mkPathString(path)))
      } else {
        List()
      }

      mismatches ++ expectedAttrs.flatMap(attr => {
        if (actualAttrs.contains(attr._1)) {
          val attrPath = appendAttribute(path, attr._1)
          val actualVal = actualAttrs.get(attr._1).get
          if (Matchers.matcherDefined("body", attrPath.asJava, matchers)) {
            logger.debug("compareText: Matcher defined for path " + attrPath)
            Matchers.domatch[BodyMismatch](matchers, "body", attrPath.asJava, attr._2, actualVal, BodyMismatchFactory.INSTANCE).asScala.toList
          } else if (attr._2 != actualVal) {
            List(new BodyMismatch(expected, actual, s"Expected ${attr._1}='${attr._2}' but received $actualVal",
              mkPathString(attrPath)))
          } else {
            List()
          }
        } else {
          List(new BodyMismatch(expected, actual, s"Expected ${attr._1}='${attr._2}' but was missing",
            mkPathString(appendAttribute(path, attr._1))))
        }
      })
    }
  }

}

File: core/matchers/src/main/kotlin/au/com/dius/pact/core/matchers/XmlBodyMatcher.kt
package au.com.dius.pact.core.matchers

import au.com.dius.pact.core.matchers.util.padTo
import au.com.dius.pact.core.model.OptionalBody
import au.com.dius.pact.core.model.matchingrules.MatchingRules
import au.com.dius.pact.core.support.zipAll
import mu.KLogging
import org.apache.xerces.dom.TextImpl
import org.w3c.dom.NamedNodeMap
import org.w3c.dom.Node
import org.w3c.dom.Node.CDATA_SECTION_NODE
import org.w3c.dom.Node.ELEMENT_NODE
import org.w3c.dom.Node.TEXT_NODE
import org.w3c.dom.NodeList
import org.xml.sax.InputSource
import java.io.StringReader
import javax.xml.XMLConstants
import javax.xml.parsers.DocumentBuilderFactory

object XmlBodyMatcher : BodyMatcher, KLogging() {

  override fun matchBody(
    expected: OptionalBody,
    actual: OptionalBody,
    allowUnexpectedKeys: Boolean,
    matchingRules: MatchingRules
  ): List<BodyMismatch> {
    return when {
      expected.isMissing() -> emptyList()
      expected.isEmpty() && actual.isEmpty() -> emptyList()
      actual.isMissing() ->
        listOf(BodyMismatch(expected.unwrap(), null, "Expected body '${expected.value}' but was missing"))
      else -> {
        compareNode(listOf("$"), parse(expected.valueAsString()), parse(actual.valueAsString()),
          allowUnexpectedKeys, matchingRules)
      }
    }
  }

  fun parse(xmlData: String): Node {
    return if (xmlData.isEmpty()) {
      TextImpl()
    } else {
      val dbFactory = DocumentBuilderFactory.newInstance()
      if (System.getProperty("pact.matching.xml.validating") == "false") {
        dbFactory.isValidating = false
        dbFactory.setFeature("http://apache.org/xml/features/nonvalidating/load-dtd-grammar", false)
        dbFactory.setFeature("http://apache.org/xml/features/nonvalidating/load-external-dtd", false)
      }
      if (System.getProperty("pact.matching.xml.namespace-aware") != "false") {
        dbFactory.isNamespaceAware = true
      }
      val dBuilder = dbFactory.newDocumentBuilder()
      val xmlInput = InputSource(StringReader(xmlData))
      val doc = dBuilder.parse(xmlInput)
      doc.documentElement
    }
  }

  private fun appendAttribute(path: List<String>, attribute: QualifiedName): List<String> {
    return path + "@${attribute.nodeName}"
  }

  fun compareText(
    path: List<String>,
    expected: Node,
    actual: Node,
    matchers: MatchingRules
  ): List<BodyMismatch> {
    val textpath = path + "#text"
    val expectedText = asList(expected.childNodes).filter { n ->
      n.nodeType == TEXT_NODE || n.nodeType == CDATA_SECTION_NODE
    }.joinToString("") { n -> n.textContent.trim() }
    val actualText = asList(actual.childNodes).filter { n ->
      n.nodeType == TEXT_NODE || n.nodeType == CDATA_SECTION_NODE
    }.joinToString("") { n -> n.textContent.trim() }
    return when {
      Matchers.matcherDefined("body", textpath, matchers) -> {
        logger.debug { "compareText: Matcher defined for path $textpath" }
        Matchers.domatch(matchers, "body", textpath, expectedText, actualText, BodyMismatchFactory)
      }
      expectedText != actualText ->
        listOf(BodyMismatch(expected, actual, "Expected value '$expectedText' but received '$actualText'",
          textpath.joinToString(".")))
      else -> emptyList()
    }
  }

  private fun asList(childNodes: NodeList?): List<Node> {
    return if (childNodes == null) {
      emptyList()
    } else {
      val list = mutableListOf<Node>()
      for (i in 0 until childNodes.length) {
        list.add(childNodes.item(i))
      }
      list
    }
  }

  private fun compareNode(
    path: List<String>,
    expected: Node,
    actual: Node,
    allowUnexpectedKeys: Boolean,
    matchers: MatchingRules
  ): List<BodyMismatch> {
    val nodePath = path + expected.nodeName
    val mismatches = when {
      Matchers.matcherDefined("body", nodePath, matchers) -> {
        logger.debug { "compareNode: Matcher defined for path $nodePath" }
        Matchers.domatch(matchers, "body", nodePath, expected, actual, BodyMismatchFactory)
      }
      else -> {
        val actualName = QualifiedName(actual)
        val expectedName = QualifiedName(expected)

        when {
          actualName != expectedName -> listOf(BodyMismatch(expected, actual,
                  "Expected element $expectedName but received $actualName",
                  nodePath.joinToString(".")))
          else -> emptyList()
        }
      }
    }

    return if (mismatches.isEmpty()) {
      compareAttributes(nodePath, expected, actual, allowUnexpectedKeys, matchers) +
        compareChildren(nodePath, expected, actual, allowUnexpectedKeys, matchers) +
        compareText(nodePath, expected, actual, matchers)
    } else {
      mismatches
    }
  }

  private fun compareChildren(
    path: List<String>,
    expected: Node,
    actual: Node,
    allowUnexpectedKeys: Boolean,
    matchers: MatchingRules
  ): List<BodyMismatch> {
    var expectedChildren = asList(expected.childNodes).filter { n -> n.nodeType == ELEMENT_NODE }
    val actualChildren = asList(actual.childNodes).filter { n -> n.nodeType == ELEMENT_NODE }
    val mismatches = if (Matchers.matcherDefined("body", path, matchers)) {
      if (expectedChildren.isNotEmpty()) expectedChildren = expectedChildren.padTo(actualChildren.size, expectedChildren.first())
      emptyList()
    } else if (expectedChildren.isEmpty() && actualChildren.isNotEmpty() && !allowUnexpectedKeys) {
        listOf(BodyMismatch(expected, actual,
          "Expected an empty List but received ${actualChildren.size} child nodes",
          path.joinToString(".")))
    } else {
      emptyList()
    }

    val actualChildrenByQName = actualChildren.groupBy { QualifiedName(it) }
    return mismatches + expectedChildren
      .groupBy { QualifiedName(it) }
      .flatMap { e ->
        if (actualChildrenByQName.contains(e.key)) {
          e.value.zipAll(actualChildrenByQName.getValue(e.key)).mapIndexed { index, comp ->
            val expectedNode = comp.first
            val actualNode = comp.second
            when {
              expectedNode == null -> emptyList()
              actualNode == null -> listOf(BodyMismatch(expected, actual,
                "Expected child <${e.key}/> but was missing",
                (path + expectedNode.nodeName + index.toString()).joinToString(".")))
              else -> compareNode(path, expectedNode, actualNode, allowUnexpectedKeys, matchers)
            }
          }.flatten()
        } else {
          listOf(BodyMismatch(expected, actual,
            "Expected child <${e.key}/> but was missing", path.joinToString(".")))
        }
      }
  }

  private fun compareAttributes(
    path: List<String>,
    expected: Node,
    actual: Node,
    allowUnexpectedKeys: Boolean,
    matchers: MatchingRules
  ): List<BodyMismatch> {
    val expectedAttrs = attributesToMap(expected.attributes)
    val actualAttrs = attributesToMap(actual.attributes)

    return if (expectedAttrs.isEmpty() && actualAttrs.isNotEmpty() && !allowUnexpectedKeys) {
      listOf(BodyMismatch(expected, actual,
        "Expected a Tag with at least ${expectedAttrs.size} attributes but received ${actual.attributes.length} attributes",
        path.joinToString(".")))
    } else {
      val mismatches = if (expectedAttrs.size > actualAttrs.size) {
        listOf(BodyMismatch(expected, actual, "Expected a Tag with at least ${expected.attributes.length} attributes but received ${actual.attributes.length} attributes",
          path.joinToString(".")))
      } else if (!allowUnexpectedKeys && expectedAttrs.size != actualAttrs.size) {
        listOf(BodyMismatch(expected, actual, "Expected a Tag with ${expected.attributes.length} attributes but received ${actual.attributes.length} attributes",
          path.joinToString(".")))
      } else {
        emptyList()
      }

      mismatches + expectedAttrs.flatMap { attr ->
        if (actualAttrs.contains(attr.key)) {
          val attrPath = appendAttribute(path, attr.key)
          val actualVal = actualAttrs[attr.key]
          when {
            Matchers.matcherDefined("body", attrPath, matchers) -> {
              logger.debug { "compareText: Matcher defined for path $attrPath" }
              Matchers.domatch(matchers, "body", attrPath, attr.value, actualVal, BodyMismatchFactory)
            }
            attr.value != actualVal ->
              listOf(BodyMismatch(expected, actual, "Expected ${attr.key}='${attr.value}' but received $actualVal",
                attrPath.joinToString(".")))
            else -> emptyList()
          }
        } else {
          listOf(BodyMismatch(expected, actual, "Expected ${attr.key}='${attr.value}' but was missing",
            appendAttribute(path, attr.key).joinToString(".")))
        }
      }
    }
  }

  private fun attributesToMap(attributes: NamedNodeMap?): Map<QualifiedName, String> {
    return if (attributes == null) {
      emptyMap()
    } else {
      (0 until attributes.length)
              .map { attributes.item(it) }
              .filter { it.namespaceURI != XMLConstants.XMLNS_ATTRIBUTE_NS_URI }
              .map { QualifiedName(it) to it.nodeValue }
              .toMap()
    }
  }
}
